Archlinux support

drinkcat opened the pull request:
This adds support for Arch Linux/Arch Linux ARM. It has been around in my repo for a long time now, with a good number of users I believe, and rarely has issues.

First 3 commits are not directly related to arch, but necessary (especially the last 2).

drinkcat committed:
arch: installer

drinkcat committed:
arch: enter-chroot: source /etc/profile.

Workaround for bug in su/bash, where /etc/profile is not sourced,
even though su is called with '-' (login shell).
See: https://bugs.archlinux.org/task/31831
This is necessary to get /usr/local/bin in the PATH, and setup the
correct locale, among other things. This is not necessary in Ubuntu,
as these are set by /etc/pam.d/su

drinkcat committed:
arch: core/post-common.

drinkcat committed:
arch: x11/xephyr/x11-common

Xorg >=1.16 needs special handling: Configure Xorg.wrap back to
its former behaviour (<1.16) with respect to starting as normal
user.

drinkcat committed:
arch: targets: Most DE and X11-related targets.

drinkcat committed:
arch: gnome

/etc/X11/xinit/xinitrc has a different meaning in Arch, and is
just a default xinit file.

drinkcat committed:
arch: chrome-common

drinkcat committed:
arch: audio and tests/30-audio

drinkcat committed:
Set UID_MIN to 1001 to make sure that useradd does not automatically add users with UID 1000.

This guarantees that UID 1000 is available when post-common adds the main user at the end of the first run of crouton.

Protect GID 1000 in the same way.

Fixes a bug when xbmc is selected during the Arch install (the xmbc user created by the package steals UID 1000).

drinkcat committed:
arch: extension: Add desktop file so it can be set as default browser.

drinkcat committed:
arch: Do not rely on /etc/os-release anymore

ArchLinux removed it from filesystem package.

drinkcat committed:
arch: kde

drinkcat committed:
tests/30-audio: Fix dd command, better pulseaudio test.

Incorrect dd direction might be causing the test to fail
occasionally (#1101).

drinkcat committed:
arch: Do not test alarm

arch and alarm are actually synonyms.

drinkcat committed:
post-common: Make sure groups actually exist

drinkcat committed:
arch: Fix group name in /etc/sudoers.d/chvt

arch uses wheel (Ubuntu/Debian uses sudo)

drinkcat committed:
arch/post-common: Fix /etc/sudoers.d/00-wheel permissions (440)

drinkcat committed:
arch/bootstrap: Add sed to the list of base packages.

auscompgeek commented on targets/post-common:
```diff
@@ -32,11 +32,37 @@ echo 'Cleaning up...' 1>&2
+        pacman --noconfirm -R $remove
+    done
+    # Clear package cache
+    yes | pacman -Scc
```
`--noconfirm`

drinkcat replied:
```
# sudo pacman -Scc --noconfirm                                                                                                                                    :(

Cache directory: /var/cache/pacman/pkg/
:: Do you want to remove ALL files from cache? [y/N] 

Database directory: /var/lib/pacman/
:: Do you want to remove unused repositories? [Y/n] 
removing unused sync repositories...
```

`--noconfirm` uses the default option, but we want `y` in both cases.

auscompgeek commented on targets/chrome-common:
```diff
@@ -26,26 +26,30 @@ REQUIRES='x11'
-    # Chrome needs hicolor-icon-theme to configure/install
-    install hicolor-icon-theme
+if [ "${DISTROAKA:-"$DISTRO"}" = 'arch' ]; then
+    install aur:google-chrome$CHANNEL
```
That's what the AUR package does.

dnschneid replied:
Yeah, and I'd rather we avoid using AUR if possible.

drinkcat replied:
There are a number of reasons to prefer AUR:
 - PKGBUILD does a number of fixups specific to Arch (https://aur.archlinux.org/packages/go/google-chrome/PKGBUILD).
 - Also, it's a PKGBUILD, so installed files are maintained in pacman database and the package can easily be removed as needed (we want to avoid "cruft": files that are not owned by any packages).
 - Finally, AUR package version is bumped after each Chrome update, so `aurcompile` can update it only when needed (on Ubuntu/Debian, Chrome adds an apt repository to address that issue).

It would not be practical to maintain our own PKGBUILD in crouton (e.g. bumping versions as needed, copy-pasting hacks from the "official" PKGBUILD), and I don't think `google-chrome` will ever get in official repository due to licensing issues. However AUR has security issues so it's not ideal either.

All in all, I'd rather drop support, instead of building something complicated that will surely break occasionally, and eat our limited resources as maintainers: After all, neither ArchLinux nor Google really supports Chrome on Arch (and, to be honest, Chrome in crouton makes little sense for most use cases). And we already do not have support on ARM, anyway...

If an Arch user really wants Chrome, it is very easy to fetch and build the PKGBUILD from AUR (the target file only does that, no crouton hack is needed!). So we push the responsibility of taking a (small) security risk onto the user.

drinkcat replied:
Fixed (dropping support.)

drinkcat replied:
Note: Instead of installing chromium automatically, just like we do on ARM (because there is no alternative), the target fails with `error 99`, so that the user can make an informed choice. (ca20031ab8ead740a7df73db97abe215ef53e7a7)

drinkcat commented on installer/arch/bootstrap:
```diff
@@ -0,0 +1,150 @@
+    curlretry -o "$FETCHDIRDB/$REPO.db" "$MIRRORBASE/$REPO.db" 1>&2
+    # Create package list in this format: $REPO:package-version
+    tar tf "$FETCHDIRDB/$REPO.db" \
+        | sed -n 's|^\(.*\)/$|'"$REPO:"'\1|p' >> "$LIST"
```
Fixed.

drinkcat replied:
Argh. `s|^\(.*\)/$|`, note the slash before the end of the line...

dnschneid replied:
Argh indeed...whoops.

drinkcat commented on installer/arch/bootstrap:
```diff
@@ -0,0 +1,150 @@
+done
+
+# Get the value of a field in the Arch package database
+# getvaluedb repo pkgver [desc|depends] field
```
Fixed.

drinkcat commented on installer/arch/bootstrap:
```diff
@@ -0,0 +1,150 @@
+                                          -e "s/\\$arch/$ARCH/"`"
+        curlretry -o "$FETCHDIRPKG/$FILE" "$MIRRORBASE/$FILE" 1>&2
+        tar --warning=no-unknown-keyword -xkf "$FETCHDIRPKG/$FILE" \
+            --exclude ".PKGINFO" --exclude ".MTREE" \
```
Oddly, it works without the `=` sign. Added.

drinkcat committed:
arch bootstrap: Simplify package installation, early keyring init

Simpler package installation and dependency resolution loop.

Better variable names (lowercase).

Initialize keyring before reinstalling the packages, so that the
signatures can be checked at that stage already.

drinkcat commented on installer/arch/bootstrap:
```diff
@@ -0,0 +1,150 @@
+
+while [ -n "$missing" ]; do
+    nextmissing=""
+    for PACKAGE in $missing; do
```
Yep. Fixed.

drinkcat commented on installer/arch/bootstrap:
```diff
@@ -0,0 +1,150 @@
+    nextmissing=""
+    for PACKAGE in $missing; do
+        # Do not install if already installed
+        for IPKG in $installed; do
```
Neat (a second * is needed after PACKAGE, though)

dnschneid replied:
I don't think so?

drinkcat replied:
You're right. Of course...

drinkcat commented on installer/arch/bootstrap:
```diff
@@ -0,0 +1,150 @@
+        done
+
+        echo "Downloading and installing $PACKAGE..." 1>&2
+        PKG="`grep ":$PACKAGE-[0-9]" "$LIST" | head -n 1`"
```
fixed.

drinkcat commented on installer/arch/bootstrap:
```diff
@@ -0,0 +1,150 @@
+            echo "Cannot find package $PACKAGE..." 1>&2
+            exit 1
+        fi
+        REPO="`echo "$PKG" | cut -f 1 -d:`"
```
Sure.

drinkcat commented on installer/arch/bootstrap:
```diff
@@ -0,0 +1,150 @@
+
+        # Get list of dependencies (ignore versions)
+        ndep="`getdbfield "$REPO" "$PKGVER" "depends" "DEPENDS" \
+                     | sed -n -e 's/^\([^<>=]*\).*$/\1/p' | tr '\n' ' '`"
```
Fixed.

drinkcat commented on installer/arch/bootstrap:
```diff
@@ -0,0 +1,150 @@
+
+echo "arch" > "$BOOTSTRAPCHROOT/etc/hostname"
+# This is only necessary for reinstall after bootstrapping: signature check
+# is eventually enabled again.
```
Yeah, that's tricky. pacman.conf is overwritten when the packages are reinstalled (I'll add a comment: it also matters for architecture specification).

Actually, I moved around the pacman keyring generation (before package reinstall), so that this is not necessary anymore.

drinkcat commented on installer/arch/getrelease.sh:
```diff
@@ -0,0 +1,46 @@
+# ArchLinux only ships an empty /etc/arch-release
+osrelease="${2%/}/etc/os-release"
+if [ -f "$osrelease" ]; then
+    rel="`sed -n -e 's/^ID=//p' "$osrelease"`"
```
Fixed.

drinkcat replied:
Fixed.

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+PKGEXT='tar.xz'
+DISTROAKA='arch'
+
+# aurcompile <package> <extra makepkg options>: compile package from AUR.
```
- `touch`: touchegg and its dependencies (frame, grail, geis). The alternative to AUR is to drop support (like we do with Debian)
 - `chrome-common`: I'll look into it. The PKGBUILD does a number of things, though, including pulling in dependencies...
 - `audio`: `lib32-sbc` (which I maintain, and is actually out of date... I wonder if we could get it merged into `[multilib]`)

dnschneid replied:
OK, could you please kick off the process of getting lib32-sbc merged into multilib?  Once that's merged I'm OK with dropping chrome and touch support for Arch.

drinkcat replied:
`sbc` is not actually needed for CRAS plugin operation (see #1353). I'll remove the function.

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+    # Expand base-devel so that install can replace packages with their
+    # multilib versions
+    install --asdeps `pacman -Sg base-devel | cut -d' ' -f 2`
+    install wget ca-certificates
```
`wget` and `ca-certificates` are already in core! Dropped the line.

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+    install wget ca-certificates
+
+    if ! id aur > /dev/null 2>&1; then
+        useradd aur -r -s /bin/false
```
I don't think it's a big deal to keep that user around (considering that the sudoers entry is deleted).

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+    su aur -s /bin/sh -c "
+        set -e
+        PACKAGE='$package'"'
+        PACKAGE2="${PACKAGE%"${PACKAGE#??}"}"
```
Ouch, this was a horrible idea... Fixed.

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+        PACKAGE='$package'"'
+        PACKAGE2="${PACKAGE%"${PACKAGE#??}"}"
+        TMPDIR="`mktemp -d --tmpdir=/tmp crouton-aur.XXX`"
+        TRAP="rm -rf '"'"'$TMPDIR'"'"'"
```
fixed

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+        trap "$TRAP" INT HUP TERM 0
+        cd "$TMPDIR"
+        wget "https://aur.archlinux.org/packages/$PACKAGE2/$PACKAGE/$PACKAGE.tar.gz"
+        tar vxf "$PACKAGE.tar.gz"
```
No.

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+        wget "https://aur.archlinux.org/packages/$PACKAGE2/$PACKAGE/$PACKAGE.tar.gz"
+        tar vxf "$PACKAGE.tar.gz"
+        cd "$PACKAGE"
+        pkgver=`sed -n '"'"'s/^pkgver=//p'"'"' PKGBUILD`
```
fixed.

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+        tar vxf "$PACKAGE.tar.gz"
+        cd "$PACKAGE"
+        pkgver=`sed -n '"'"'s/^pkgver=//p'"'"' PKGBUILD`
+        pkgrel=`sed -n '"'"'s/^pkgrel=//p'"'"' PKGBUILD`
```
Quoted.

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+        pkgrel=`sed -n '"'"'s/^pkgrel=//p'"'"' PKGBUILD`
+        # FIXME: Temporary fix for mawk (the AUR PKGBUILD currently works, but
+        # will break again on the next upstream update)
+        if [ "$PACKAGE" = "mawk" ]; then
```
The "bug report" is the my comment under `nboichat` there: https://aur.archlinux.org/packages/mawk/ .

Anyway, we do not need `mawk` anymore, dropping the hack.

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+        # will break again on the next upstream update)
+        if [ "$PACKAGE" = "mawk" ]; then
+            sed -i PKGBUILD -e '\''s|source=.*|\
+source=("ftp://invisible-island.net/mawk/$pkgname-${pkgver//_/-}.tgz")|'\''
```
dropped.

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+            echo "Package $PACKAGE already up to date ($pkgver-$pkgrel)."
+        else
+            makepkg -A -s -i '"$*"' --noconfirm
+        fi'
```
Fixed.

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+    fi
+    while [ "$#" != 0 ]; do
+        if [ "$1" = "--minimal" ]; then
+            true # Ignore: irrelevant for arch
```
Sure.

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+            shift
+            break
+        fi
+        aurstrip=${1#aur:}
```
fixed

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+        fi
+        aurstrip=${1#aur:}
+        if [ "$aurstrip" = "$1" ]; then
+            if [ -n "$multilib" ] && \
```
Fixed.

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+    fi
+    shift
+    for pkg in "$@"; do
+        if ! pacman -Q "$pkg" 2>/dev/null >/dev/null; then
```
Fixed.

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+# Make sure package architecture is set manually.
+# Using the default Auto here detects the chroot from uname -m, breaking i686
+# chroots on x86-64.
+sed -i 's/^[ \t]*Architecture[ \t].*/Architecture = '"$ARCH"'/' /etc/pacman.conf
```
Moved comment to boostrap, and removed the line here. I think this was meant to possibly recover from this issue: https://github.com/drinkcat/chroagh/issues/35. But if this ever becomes necessary, the chroot is probably beyond recovery already (some x86_64 packages were probably installed)

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+    # exact same packages as during the bootstrap creation, to make sure we
+    # do not leave any cruft behind).
+    if [ -n "$MIRROR" -a "$MIRROR" != 'unspecified' ]; then
+        echo "Server = http://www.example.com/" > "/etc/pacman.d/mirrorlist"
```
Didn't know about 192.0.2.1, nice! (example.com is also reserved, but might be hijackable by a rogue DNS server I guess)

drinkcat committed:
Other fixes after review.

drinkcat committed:
arch: chrome-common/touch: Drop support.

Prepare for removal or AUR handler.

odinuge commented:
Any news on this? 
I have not looked into all the code, but after a few weeks of testing it seems fine!
(Tested on an Acer CB5, Tegra K1)

drinkcat commented:
> Any news on this?

As you can see there are a large number of comments in the review. Then, I need to rebase on the latest crouton, and port new features like xiwi. So, it's going to take some time...

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+
+    # Reset architecture
+    sed -i 's/^[ \t]*Architecture[ \t].*/Architecture = '"$ARCH"'/' \
+        /etc/pacman.conf
```
Well, only twice now. Once in `bootstrap` and once here in `prepare`... These 2 scripts do not share a common include file, so I guess we have to live with the repetition...

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+
+    # Restore the original /etc/resolv.conf
+    if [ -f /etc/resolv.conf.pacorig ]; then
+        rm -f /etc/resolv.conf
```
Yes. Fixed.

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+    if [ "${ARCH#arm}" = "$ARCH" ]; then
+        cat 1>&2 <<END
+Initializing pacman keyring...
+If this gets stuck, generate entropy by moving the mouse around (e.g. do some
```
Sure. Fixed.

drinkcat committed:
arch/alarm: Setup keyring on alarm as well

Needs a bit more processing as it is not the default.

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+        mv -f /etc/resolv.conf.pacorig /etc/resolv.conf
+    fi
+
+    # Initialize pacman keyring (i686/x86_64 only)
```
Actually, there is. It's just that it's not enabled by default (http://archlinuxarm.org/about/package-signing).

Added support.

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+    # Packages that are in base, but that we do not want to install in the
+    # chroot (nothing prevents the user from installing them later on)
+    PACKAGES_TARGET_IGNORE="cronie device-mapper dhcpcd heirloom-mailx \
+    jfsutils logrotate lvm2 mdadm netctl pciutils reiserfsprogs \
```
Fixed.

drinkcat commented on installer/arch/prepare:
```diff
@@ -0,0 +1,244 @@
+    basepkg=""
+    for PACKAGE in $allbasepkg; do
+        inst='y'
+        for IPKG in $PACKAGES_TARGET_IGNORE; do
```
Sure!

drinkcat commented on targets/core:
```diff
@@ -8,6 +8,12 @@ CHROOTBIN='croutonversion host-dbus host-x11'
+# Make sure that useradd does not automatically add users or groups with
+# UID/GID 1000, so that it is guaranteed to be available when post-common
+# adds the main user
+sed -i -e 's/^\([GU]ID_MIN[ \t]*\)1000$/\11001/' /etc/login.defs
```
Very neat. Done!

drinkcat commented on targets/core:
```diff
@@ -102,6 +108,29 @@ EOF
+    pacman -Syu --noconfirm || ret=$?
+    if [ "$ret" != 0 ]; then
+        cat 1>&2 <<END
+Cannot upgrade the system automatically, you may want to run the installer
```
Fixed.

drinkcat commented on targets/extension:
```diff
@@ -29,12 +29,33 @@ if [ "$XMETHOD" = 'x11' ]; then
+elif [ "${DISTROAKA:-"$DISTRO"}" = 'arch' ]; then
+    install arch=desktop-file-utils,
+
+    cat > '/usr/local/share/applications/crouton.desktop' <<END
```
> Do you need to ensure the directory exists first?

Yes. Fixed.

> We probably want to do this on Debian as well.

I did it that way because `update-alternatives` does not exist on Debian. But, yeah, I guess we could use that approach on Debian too. I'll add a FIXME in the code, for now.

drinkcat commented on targets/extension:
```diff
@@ -29,12 +29,33 @@ if [ "$XMETHOD" = 'x11' ]; then
+[Desktop Entry]
+Encoding=UTF-8
+Type=Application
+Name=Crouton URL handler
```
fixed.

drinkcat commented on targets/extension:
```diff
@@ -29,12 +29,33 @@ if [ "$XMETHOD" = 'x11' ]; then
+Type=Application
+Name=Crouton URL handler
+GenericName=Web Browser
+Comment=Open URL in Chromium OS via crouton extension
```
fixed.

drinkcat commented on targets/extension:
```diff
@@ -29,12 +29,33 @@ if [ "$XMETHOD" = 'x11' ]; then
+GenericName=Web Browser
+Comment=Open URL in Chromium OS via crouton extension
+Exec=/usr/local/bin/croutonurlhandler %U
+Icon=crouton
```
No (as far as I remember). Might be worth installing the extension icon and use that.

drinkcat commented on targets/x11-common:
```diff
@@ -70,3 +75,28 @@ ResultAny=no
+    order="firefox.desktop;google-chrome.desktop;chromium.desktop"
+    order="$order;crouton.desktop;netsurf.desktop"
+    mimedefault='/usr/local/share/applications/mimeapps.list'
+    echo "[Added Associations]" > $mimedefault
```
Arguably, yes. Users would be better off using `~/.config/mimeapps.list`, though (https://wiki.archlinux.org/index.php/default_applications). I'll replace with an awk script.

drinkcat replied:
Nah. I'll do it the "classical" way: A symlink to `/etc/crouton/mimeapps.list`.

drinkcat commented on targets/xephyr:
```diff
@@ -28,13 +28,18 @@ elif release -gt quantal; then
 fi
 
 # xserver-xephyr won't auto replace the manually-downloaded version.
+# FIXME: What is the purpose of dmz-cursor-theme and libgl1-mesa-dri?
```
Okay. Didn't seem necessary on arch.

drinkcat commented on targets/xephyr:
```diff
@@ -28,13 +28,18 @@ elif release -gt quantal; then
 # Compile croutoncursor
 compile cursor '-lX11 -lXfixes -lXrender' \
-    libx11-dev libxfixes-dev libxrender-dev
+    arch=,libx11-dev arch=,libxfixes-dev arch=,libxrender-dev
```
True. But that'd mean overloading `compile()` in `installer/prepare` for each distro... (we still want to install `gcc` as dependency on arch). I'd say the current approach is still cleaner...

drinkcat commented on targets/xorg:
```diff
@@ -91,9 +91,20 @@ END
+fi
+
+# Xorg 1.16 wants to run as normal user, which breaks crouton
+if [ "${DISTROAKA:-"$DISTRO"}" = 'arch' ]; then
```
sure.

drinkcat commented on targets/xorg:
```diff
@@ -91,9 +91,20 @@ END
+
+# Xorg 1.16 wants to run as normal user, which breaks crouton
+if [ "${DISTROAKA:-"$DISTRO"}" = 'arch' ]; then
+    cat >/etc/X11/Xwrapper.config <<EOF
```
Arch doesn't ship with a default file.

drinkcat commented on test/tests/30-audio:
```diff
@@ -11,6 +11,23 @@ if [ -z "$release" ]; then
 
 snapshot "$release" core
+
+# FIXME: We are lacking a test for arch non-multilib here...
```
Added a simple test (no pulseaudio): it's mostly the crouton installer that I want to test.

drinkcat committed:
Another set of fixes after review.

drinkcat commented:
Addressed everything but:
- https://github.com/dnschneid/crouton/pull/1201#discussion_r23336865
- https://github.com/dnschneid/crouton/pull/1201#discussion_r23338406

Will proceed to rebase...

drinkcat commented:
Note to self, I think install with a proxy is broken (some fetching is done before the proxy is setup...)

drinkcat committed:
Merge remote-tracking branch 'crouton/master' into croagh_arch

Only resolving clear conflicts.

Resolved conflicts:
	targets/audio
	targets/chrome-common
	targets/core
	targets/touch
	targets/xorg
	test/tests/30-audio

drinkcat committed:
test/tests/35-xorg,w0-common: Arch support.

drinkcat committed:
xiwi: Add Arch support.

drinkcat committed:
targets/*-desktop: Arch support.

Sometimes approximate (probably not as "full" as ubuntu desktop
packages), and cannot find anything to add for lxde-desktop.

drinkcat committed:
targets/core: No apt-transport on Arch.

drinkcat committed:
xiwi: arch, install xorg-server-devel

drinkcat committed:
Arch: Remove aurcompile, not needed anymore.

drinkcat committed:
xserverrc-xiwi: Prefer Xorg.bin, then Xorg.

On Arch, Xorg is setuid, and Xorg.bin is non-setuid, while on
Ubuntu/Debian, X is setuid, and Xorg is non-setuid (Xorg.bin does
not exist). So prefer Xorg.bin, then fall back on Xorg.

drinkcat committed:
Merge remote-tracking branch 'crouton/master' into croagh_arch.tmp

Conflicts:
	host-bin/enter-chroot

drinkcat committed:
xserverrc-xiwi: Better detection of appropriate Xorg executable

drinkcat commented:
Merged master back into this branch, added xiwi support.

Still a number of issues:
 - The 2 comments above
 - Proxy support is broken
 - Error message on pacman update is broken (`sudo enter-chroot -n  -u 0`)
 - Test xbmc
 - Add hold/unhold function to reduce code duplication, adapt to xiwi target

drinkcat committed:
Merge remote-tracking branch 'crouton/master' into croagh_arch

mathuin commented:
Today I checked out this branch and ran 'make all'.  The binary `croutonfreon` failed to build, apparently because `src/freon.c` is missing a `main()` function.  I'm insufficiently familiar with crouton to know what that really means.  The `crouton` shell script built without errors.

I ran the script with the following command line:  `sudo sh -e ./Downloads/croagh_arch -r arch -t xiwi,cli-extra`.  It exited with the following error:
```
Patching Xorg dummy driver (xrandr 1.2 support)...
patching file src/dummy_driver.c
Compiling Xorg dummy driver...
Installing Xorg dummy driver...
/prepare.sh: line 5035: syntax error near unexpected token `<<<'
Failed to complete chroot setup.
Unmounting /usr/local/chroots/arch...
chronos@localhost ~ $ 
```
I tracked this down to `targets/xiwi` having an unresolved merge between `HEAD` and `crouton/master`, probably during a rebase.  A quick `git grep` shows that this is the only merge conflict.  I chose the `HEAD` version, rebuilt `crouton`, deleted my incomplete chroot, and started again.  This time it ran through to completion.

The one thing I don't like is the removal of certain packages like binutils at the end of the build.  I don't feel strongly enough to defend it at the moment, though, but if I do I'll submit a pull request.

mathuin committed:
Minor fix to broken merge.

At some point when HEAD was rebased or merged with crouton/master, one
conflict went unresolved.  I resolved the conflict and was able to use
crouton to successfully generate an arch chroot.

dnschneid commented:
@mathuin also pointed out that xterm isn't installed automatically on some targets.

Commit:
Merge pull request #81 from mathuin/fix-xiwi-merge-issue

Minor fix to broken merge.

EddieRingle commented:
FWIW I have this PR and the i3xiwi branch applied and so far have been having a fairly good experience. Can even bring up Android Studio in a xiwi window, which is absolutely awesome.

@dnschneid I dug around earlier before doing this and noticed the xiwi-app branch had been killed and I couldn't find any explanation as to why. Are those changes now a part of the xiwif branch or something? Perhaps that's one more set of changes I should apply.

dnschneid commented:
@EddieRingle xiwi-app got merged into master.  i3xiwi is likely to get merged next, and xiwif is a bit of a testing ground at the moment.

EddieRingle commented:
@dnschneid Good to know! And yeah, the i3xiwi branch is running pretty smooth here.

mathuin committed:
Added ltspackages variable to dummy video.

I mismerged a reference to xserver-xorg-video-dummy and forgot to
change it to xserver-xorg-video-dummy$ltspackages as per another
update.  This commit fixes that mistake.

Commit:
Merge pull request #84 from mathuin/bad-dummy-forgot-ltspackages

Added ltspackages variable to dummy video.

drinkcat committed:
Merge remote-tracking branch 'crouton/master' into croagh_arch

Conflicts: (all minor nits)
	targets/audio
	targets/gtk-extra
	targets/xfce-desktop
	targets/xiwi

drinkcat committed:
arch: Add support install_dummy_dist

By support, I mean, no support... I don't think we need this kind
of stuff on Arch. At least not for now...

drinkcat committed:
arch: xiwi: install i3-wm instead of i3

We don't need i3status or i3lock.

drinkcat commented:
Merged crouton/master back into this branch. Issues here still stand: https://github.com/dnschneid/crouton/pull/1201#issuecomment-75229301 .

And XMETHOD=xorg on freon is broken (you can start a server, but input devices do not to get grabbed, and you cannot switch back to Chrome OS).

MichaelMackus commented:
Any chance for a rebase @drinkcat ?

Thanks for the work on this!

MattSturgeon commented:
It looks like a lot of effort has gone into this, is it still being worked on?

salinasv commented:
@drinkcat are you still working on this?
Can you point us out what is missing? maybe some of us can take this up to a merge ready state.

dnschneid commented:
Lots of maintenance issues here.  Maybe we can simplify things by using one of the [pre-built bootstraps](https://mirrors.kernel.org/archlinux/iso/latest/)?  I don't know how heavy they are; ARM would have to use something like [a system image](http://os.archlinuxarm.org/os/ArchLinuxARM-armv7-latest.tar.gz), but that would probably install too much.

dnschneid commented:
Also, might want to consider not having separate arch/alarm releases

eyqs commented:
What is the current status of this pull request?

eyqs commented:
What is the current status of this pull request? Are you still actively working on it, @drinkcat?

zhaofengli commented:
@qntmpkts maintains a fork at https://github.com/qntmpkts/chroagh, and it works fine. To make it easier to keep up with upstream, I rebased all changes on that repo (except stuff no longer supported by upstream crouton) against crouton at [zhaofengli/crouton](https://github.com/zhaofengli/crouton). All commits there should cleanly apply.

EnderNightLord-ChromeBook commented:
Drinkcat Have you Updated chrough Yet it's been since 2013 since a release has came out and it does not work anymore!!!
